packages/block-editor/src/components/block-toolbar/index.js

```diff
@@ -239,6 +234,11 @@ export function PrivateBlockToolbar( {
  * @param {string}  props.variant        Style variant of the toolbar, also passed to the Dropdowns rendered from Block Toolbar Buttons.
  */
 export default function BlockToolbar( { hideDragHandle, variant } ) {
+	const hasBlockToolbar = useHasBlockToolbar();
+	if ( ! hasBlockToolbar ) {
+		return null;
+	}
```

Comment:
Moving this check to before the `<PrivateBlockToolbar />` avoids making an unnecessary `useSelect`.